RoyaltiesProcessorDelivery/src/services/opensea_sales_service.ts

```diff
@@ -84,14 +145,137 @@ export class OpenSeaSalesService {
+      }
+      // warning to devs in future - when adding ability to scroll, keep in mind
+      // that TheGraph has an upper limit on skip, so may need to filter in
+      // some other way than just using projects() query if future-proofing.
+    }
+    console.log("");
+    // query OpenSea api for every token zero to build array of collection slugs
+    // iterate one-by-one to eliminate too-many-calls response from OS API
+    // use local cache because this takes a long time due to OS's API.
+    const slugsAndTokenZeros: T_SlugAndTokenZero[] = [];
+    for (let i = 0; i < tokenZeros.length; i++) {
+      const _tokenZero = tokenZeros[i];
+      console.debug(
+        `[INFO] Getting OS collection slug for: ${_tokenZero.tokens[0].project.name}`
+      );
+      let collectionSlug = collectionSlugCache.getKey(
+        _tokenZero.tokens[0].project.id
+      );
+      if (collectionSlug !== undefined) {
+        console.info(`[INFO] using cached collection slug ${collectionSlug}`);
+        slugsAndTokenZeros.push({
+          collectionSlug: collectionSlug,
+          tokenZero: _tokenZero,
+        });
+      } else {
+        try {
+          collectionSlug = await getOpenSeaAssetCollectionSlug(
+            _tokenZero.tokens[0].contract.id,
+            _tokenZero.tokens[0].tokenId.toString()
+          );
+        } catch (error) {
+          // likely too many requests, cool off for ten seconds
+          console.warn(
+            "[WARNING] likely too many requests... cooling down for 20 seconds"
+          );
+          await delay(15000);
+          // OS API appears to make us fail once more before working
+          try {
+            await getOpenSeaAssetCollectionSlug(
+              _tokenZero.tokens[0].contract.id,
+              _tokenZero.tokens[0].tokenId.toString()
+            );
+          } catch {
+            console.log("[INFO] failed once more as expected");
+          }
+          await delay(5000);
+          console.info("[INFO] resuming");
+          collectionSlug = await getOpenSeaAssetCollectionSlug(
+            _tokenZero.tokens[0].contract.id,
+            _tokenZero.tokens[0].tokenId.toString()
+          );
+          console.debug("[INFO] successfully recovered");
+        }
+        slugsAndTokenZeros.push({
+          collectionSlug: collectionSlug,
+          tokenZero: _tokenZero,
+        });
+        // add slug to cache, save
+        collectionSlugCache.setKey(
+          _tokenZero.tokens[0].project.id,
+          collectionSlug
+        );
+        collectionSlugCache.save(true);
+        // Throttle due to OpenSea API rate limiting
+        await delay(500);
+      }
+    }
+    // OS api works in terms of timestamps, not blocks.
+    let maxTimestamp = await getBlockTimestamp(blockRange[1]);
+
+    // retrieve all events in timestamp/block range, for each collection
+    // populate openSeaSales along the way!
+    for (let i = 0; i < Math.min(99999999, slugsAndTokenZeros.length); i++) {

```

Comment:
I think we should document here the usage of `99999999`

Reply:
oh yeah, oops, removing this. I think was using this to debug some stuff early on!